File: Python/Medium/1262_Greatest_Sum_Divisible_by_Three/solution.py
from typing import List

class Solution:
    def maxSumDivThree(self, nums: List[int]) -> int:
        return Solution.helper(nums, 0, 0, {})

    def helper(nums, p, rem, memo):
        key = (p, rem)
        if key in memo:
            return memo[key]

        if p == len(nums):
            if rem == 0:
                return 0
            return float('-inf')
        
        pick = nums[p] + Solution.helper(nums, p+1,(rem + nums[p]) % 3, memo)
        dunPick = Solution.helper(nums, p+1, rem, memo)

        memo[key] = max(pick, dunPick)
        return memo[key]

File: Golang/Medium/1262_Greatest_Sum_Divisible_by_Three/solution.go
package greatestsumdivisiblebythree

import "math"

type state struct {
	p   int
	rem int
}

func maxSumDivThree(nums []int) int {
	return helper(nums, 0, 0, make(map[state]int))
}

func helper(nums []int, p int, rem int, memo map[state]int) int {
	key := state{p, rem}
	if v, exist := memo[key]; exist {
		return v
	}

	if p == len(nums) {
		if rem == 0 {
			return 0
		}
		return int(math.Inf(-1))
	}

	pick := nums[p] + helper(nums, p+1, (rem+nums[p])%3, memo)
	dunPick := helper(nums, p+1, rem, memo)
	memo[key] = max(pick, dunPick)
	return memo[key]
}
